### PR TITLE
[MO] - [system test] -> test suite of multiple listeners

### PR DESCRIPTION
**Pull request by see-quick**
Signed-off-by: morsak <xorsak02@stud.fit.vutbr.cz>

### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR introduces a new system suite for a related feature of https://github.com/strimzi/strimzi-kafka-operator/pull/3603. 

#### Test cases

Here is so abstract overview of the test cases which I have designed. Description is not needed because names are self-explanatory:

- testMultipleNodePorts 
- testMultipleInternal
- testCombinationOfInternalAndExternalListeners
- testMultipleLoadBalancers
- testMultipleRoutes
- testMixtureOfExternalListeners
- testCombinationOfEveryKindOfListener

Time Consumption = [TODO] minutes

### Checklist

- [x] Write tests
- [x] Make sure all tests pass

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                            .withTls(stochasticCommunication)
+                            .build());
+                    }
+                    break;
+                case ROUTE:
+                    stochasticCount = ThreadLocalRandom.current().nextInt(2, 3);
+
+                    for (int j = 0; j < stochasticCount; j++) {
+
+                        boolean stochasticCommunication = ThreadLocalRandom.current().nextInt(2) == 0;
```
Comment (see-quick) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:213]: this will be removed because `Route or Ingress type listener and requires enabled TLS encryption`

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaResources.java [hunk 2]
```diff
@@ -138,6 +138,19 @@ public static String tlsBootstrapAddress(String clusterName) {
...
+     * @param clusterName  The {@code metadata.name} of the {@code Kafka} resource.
+     * @param port The {@code spec.kafka.listeners.port} of the {@code Kafka} resource.
+     * @return The address of the corresponding bootstrap {@code Service}.
+     * @see #plainBootstrapAddress(String)
+     * @see #tlsBootstrapAddress(String)
+     */
+    public static String bootstrapAddressOnSpecificPort(String clusterName, int port) {
+        return bootstrapServiceName(clusterName) + ":" + port;
+    }
+
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaResources.java:153]: I'm not sure I think we really need this here.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 3]
```diff
@@ -142,48 +170,67 @@ public void verifyProducedAndConsumedMessages(int producedMessages, int consumed
      * Get external bootstrap connection
      * @param namespace kafka namespace
      * @param clusterName kafka cluster name
+     * @param listenerName name of the listener
      * @return bootstrap url as string
      */
-    public static String getExternalBootstrapConnect(String namespace, String clusterName) {
+    @SuppressWarnings("Regexp") // because of extBootstrapService.getSpec().getType().toLowerCase()
+    @SuppressFBWarnings("DM_CONVERT_CASE")
+    public static String getExternalBootstrapConnect(String namespace, String clusterName, String listenerName) {
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:178]: When you deploy Kafka, the bootstrap addresses for all listeners get filled in the status section. Is there any actual reason why you need to reverse engineer the address here instead of reading it simply from the status? All you would need is a listener name, you would not need to both with OpenShfit, service types etc.

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java [hunk 4]
```diff
@@ -169,7 +169,7 @@ private static KafkaBuilder defaultKafka(Kafka kafka, String name, int kafkaRepl
                         .endGenericKafkaListener()
                     .endListeners()
                     .withNewInlineLogging()
-                        .addToLoggers("log4j.rootLogger", "DEBUG")
+                        .addToLoggers("log4j.rootLogger", "DEBUG, CONSOLE")
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:172]: Aren't we removing this in the other PRs? Please check with @stanlyDoge and @mstruk. It is IIRC also not what the docs suggest ... we should primarily follow our docs to make sure the docs are valid and test them this way as well.

File: systemtest/pom.xml [hunk 5]
```diff
@@ -38,14 +38,6 @@
...
             <artifactId>log4j-api</artifactId>
         </dependency>
-        <dependency>
-            <groupId>org.apache.logging.log4j</groupId>
-            <artifactId>log4j-core</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.logging.log4j</groupId>
-            <artifactId>log4j-slf4j-impl</artifactId>
-        </dependency>
```
Comment (see-quick) [inline systemtest/pom.xml:48]: Removing because these dependencies are un-used in `systemtest` module

**Review by scholzj — COMMENTED** [review]
The test looks IMHO good. But I left some comments around the support tooling.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java [hunk 6]
```diff
@@ -81,6 +82,29 @@ protected Builder self() {
...
+        builder.withProducerProperties(oauthExternalKafkaClient.getProducerProperties());
+        builder.withConsumerProperties(oauthExternalKafkaClient.getConsumerProperties());
+        builder.withOauthClientId(oauthExternalKafkaClient.getClientId());
+        builder.withClientSecretName(oauthExternalKafkaClient.getClientSecretName());
+        builder.withOauthTokenEndpointUri(oauthExternalKafkaClient.getOauthTokenEndpointUri());
+        builder.withIntrospectionEndpointUri(oauthExternalKafkaClient.getIntrospectionEndpointUri());
+
+        return builder;
+    }
+
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java:107]: I guess we discussed this when I was working on the example clients builder but -> isn't there some way how to just set the "properties" that are specific for Oauth external clients here and the others set in the Abstract? I know that there can be some dependency (etc.) issues, but still asking...

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java (hunk 4 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:172]: Leftover, when I was debugging some issues.

Comment (im-konge) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:172]: Like @scholzj mentioned, I also guess that this is part of another PR, I don't know if @Frawless didn't do something similar.

Comment (stanlyDoge) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:172]: PR fixing this is not merged yet https://github.com/strimzi/strimzi-kafka-operator/pull/3672

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        multipleDifferentListeners.addAll(internalListeners);
+        multipleDifferentListeners.addAll(nodeportListeners);
+        multipleDifferentListeners.addAll(routeListeners);
+        multipleDifferentListeners.addAll(loadbalancersListeners);
+
+        // run INTERNAL + NODEPORT + ROUTE + LOADBALANCER listeners
+        runTestCase(multipleDifferentListeners);
+    }
+
+    private void runTestCase(List<GenericKafkaListener> listeners) {
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:135]: Maybe different name?

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+        LOGGER.info("Starting to generate test cases for multiple listeners");
+
+        int stochasticCount;
+
+        for (KafkaListenerType kafkaListenerType : KafkaListenerType.values()) {
+
+            LOGGER.info("Generating {} listener", kafkaListenerType.name());
+
+            List<GenericKafkaListener> testCase = new ArrayList<>(5);
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:257]:
```suggestion
            List<GenericKafkaListener> testCaseListeners = new ArrayList<>(5);
```?

**Review by im-konge — COMMENTED** [review]
This looks good, you've done really good work (as always). Just some suggestions/questions like always :)

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java (hunk 7 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:135]: Maybe suggest something?

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 6 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java]: Thanks for catch, I have re-designed it a little bit and now in the client will not be a redundant code.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaResources.java (hunk 2 above)
Comment (see-quick) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaResources.java:153]: Ok I will remove it.

**Comment by see-quick** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MultipleListenersST

**Comment by see-quick** [comment]
It seems that Jenkins does not correctly count the results. 7 tests were executed and 3 fails. Anyway, there is some problem with DNS resolving when we use routes.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java (hunk 6 above)
Comment (im-konge) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/externalClients/OauthExternalKafkaClient.java]: Thanks a lot :)

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java (hunk 7 above)
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:137]:
```suggestion
    private void runMultipleListenersTest(List<GenericKafkaListener> listeners) {
```
or
```suggestion
    private void runListenersTest(List<GenericKafkaListener> listeners) {
```

**Comment by Frawless** [comment]
@see-quick Could you please compare how route address looks in your scenarios and scenaros which are working fine (ListenersST) ?

**Comment by see-quick** [comment]
@strimzi-ci run tests test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MultipleListenersST

**Comment by Frawless** [comment]
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Flakes: 1

**Review by scholzj — APPROVED** [review]

**Review by im-konge — APPROVED** [review]
LGTM, good job :)

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ConsumerProperties.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            return this;
+        }
+
+        public ConsumerPropertiesBuilder withClientIdConfig(String clientId) {
+
+            this.properties.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, clientId);
+            return this;
+        }
+
+        @SuppressWarnings("Regexp") // for the `.toLowerCase()` because kafka needs this property as lower-case
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ConsumerProperties.java:52]: Why is this needed? You should use it like this ` .toLowerCase(Locale.ENGLISH)` I think you can remove `DM_CONVERT_CASE` as well after the change.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ConsumerProperties.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    private ConsumerProperties(ConsumerPropertiesBuilder builder) {
+        super(builder);
+        properties = builder.properties;
+    }
+
+    @Override
+    @SuppressWarnings({"Regexp", "unchecked"}) // for the `.toUpperCase()` because OffsetStrategy needs to be upper-case
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ConsumerProperties.java:77]: Same here

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ProducerProperties.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        protected ProducerPropertiesBuilder self() {
+            return this;
+        }
+    }
+
+    private ProducerProperties(ProducerPropertiesBuilder builder) {
+        super(builder);
+
+        if (builder.properties.getProperty(ProducerConfig.MAX_BLOCK_MS_CONFIG) == null || builder.properties.getProperty(ProducerConfig.MAX_BLOCK_MS_CONFIG).isEmpty()) {
+            LOGGER.info("Setting default value of {} to {}", ProducerConfig.MAX_BLOCK_MS_CONFIG, DEFAULT_MAX_BLOG_MS_CONFIG);
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/clientproperties/ProducerProperties.java:73]: debug?

**Review by stanlyDoge — APPROVED** [review]

**Comment by see-quick** [comment]
@strimzi-ci run tests test_only profile=externalclients,internalclients

**Review by scholzj — APPROVED** [review]

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java [hunk 12]
```diff
@@ -4,21 +4,20 @@
...
 import org.apache.logging.log4j.Logger;
 
 import java.security.InvalidParameterException;
+import java.util.List;
 
-import static io.strimzi.api.kafka.model.KafkaResources.externalBootstrapServiceName;
-import static io.strimzi.test.k8s.KubeClusterResource.kubeClient;
-
-public abstract class AbstractKafkaClient {
+@SuppressWarnings("unchecked")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:19]: I'm guessing this is because the fluent pattern on the builder with each method having to cast `(SELF) this` results in lots of warnings. It is best to use `@SuppressWarnings` on the smallest scope possible, so that genuine warnings are not suppressed accidentally. But you understandably don't want to have to annotate each method, so the solution is to write a method `self()` does the cast and is annotated, then all the rest get to `return self();` and not needing annotating. Basically the rule of thumb is always to refactor you untypesafe code into a method and annotate just that method.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+
+    @OpenShiftOnly
+    @Tag(EXTERNAL_CLIENTS_USED)
+    @Test
+    void testMultipleRoutes() {
+        runListenersTest(testCases.get(KafkaListenerType.ROUTE));
+    }
+
+    @Tag(NODEPORT_SUPPORTED)
+    @OpenShiftOnly
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:99]: Probably best to have a consistent order for the annotations, and not split annotations of the same type.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+        for (GenericKafkaListener listener : listeners) {
+
+            String topicName = KafkaTopicUtils.generateRandomNameOfTopic();
+            KafkaTopicResource.topic(CLUSTER_NAME, topicName).done();
+
+            boolean isTlsEnabled = listener.isTls();
+
+            if (listener.getType() != KafkaListenerType.INTERNAL) {
+
+                if (isTlsEnabled) {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:166]: The two branches of this if look very similar.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+                        testCaseListeners.add(new GenericKafkaListenerBuilder()
+                            .withName(generateRandomListenerName())
+                            .withPort(11900 + j)
+                            .withType(KafkaListenerType.LOADBALANCER)
+                            .withTls(stochasticCommunication)
+                            .build());
+                    }
+                    break;
+                case ROUTE:
+                    testCaseListeners.add(new GenericKafkaListenerBuilder()
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:300]: Why not a random number of routes?

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+                    internalPlainKafkaClient.checkProducedAndConsumedMessages(
+                        internalPlainKafkaClient.sendMessagesPlain(),
+                        internalPlainKafkaClient.receiveMessagesPlain()
+                    );
+                }
+            }
+        }
+    }
+
+    private Map<KafkaListenerType, List<GenericKafkaListener>> generateTestCases() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:256]: Deserves some javadoc to explain what this is doing.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+                            .withPort(13900 + j)
+                            .withType(KafkaListenerType.INTERNAL)
+                            .withTls(stochasticCommunication)
+                            .build());
+                    }
+            }
+            testCases.put(kafkaListenerType, testCaseListeners);
+        }
+
+        LOGGER.info("Finished will generation of test cases for multiple listeners");
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:326]:
```suggestion
        LOGGER.info("Finished with generation of test cases for multiple listeners");
```
and maybe log the listeners generated, if it's not logged elsewhere?

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+
+        return testCases;
+    }
+
+    private String generateRandomListenerName() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:331]: You could perhaps reuse the `PasswordGenerator` we have somewhere?

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+
+        for (KafkaListenerType kafkaListenerType : KafkaListenerType.values()) {
+
+            LOGGER.info("Generating {} listener", kafkaListenerType.name());
+
+            List<GenericKafkaListener> testCaseListeners = new ArrayList<>(5);
+
+            switch (kafkaListenerType) {
+                case NODEPORT:
+                    stochasticCount = ThreadLocalRandom.current().nextInt(2, 5);
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:270]: The problem with random tests is that they're not reproducible, so if you get a failure and think you've fixed it how do you re-run the same test again? The solution is in the fact that `java.util.Random` is a pseudo-random number generator and the contract says if you give it the same seed it will generate the same sequence. But got can't get the seed. So in normal running you can generate a seed using one `Random` (or use `System.nanoTime()` or whatever) and use it as the seed for another Random instance and _log the seed_. Then you can reproduce any failed test for which you have the logs by having some System property or env var which allows you to set the seed, rather than using a `Random`/`System.nanoTime()` whatever.

If we're going to use more random testing in the future this is a pattern we should encapsulate, but in another PR.

**Review by tombentley — APPROVED** [review]
A few nits, but LGTM.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java (hunk 12 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java:19]: Yeah, make sense thanks.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java (hunk 15 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:300]: That's an issue with the multiple routes we spoke with @scholzj (related with DNS) I don't know it is already resolved.

Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/listeners/MultipleListenersST.java:300]: It is not related to DNS - this was the bug with unique ports per listener which should be fixed now in Kafka 2.7.0 - you should leave there some comment / TODO to add it later.

**Review by Frawless — APPROVED** [review]

**Comment by see-quick** [comment]
This Jenkins result is old. In Azure latest results were with 5 errors. I already fixed 3 of them. There are only 2 left
```
1. testUserInDifferentNamespace
2. testRecoveryDuringZookeeperRollingUpdate
```
and logs does not seem that my change invoked these errors. In case we want to merge this as soon as possible. I think it's ready. I will investigate this problem and then create PR with fix of these 2 test cases.

**Comment by scholzj** [comment]
Does not look good ...

```
[ERROR] Errors: 
[ERROR] io.strimzi.systemtest.kafka.listeners.ListenersST.testCustomChainCertificatesForNodePort
[ERROR]   Run 1: ListenersST.testCustomChainCertificatesForNodePort:701 NullPointer
[ERROR]   Run 2: ListenersST.testCustomChainCertificatesForNodePort:701 NullPointer
[ERROR]   Run 3: ListenersST.testCustomChainCertificatesForNodePort:701 NullPointer
```

**Comment by see-quick** [comment]
The last test is fixed and verified on my local machine.

**Review by Frawless — APPROVED** [review]
